Sim/SimDelphesInterface/src/DelphesSimulation.cpp

```diff
@@ -343,78 +369,452 @@ StatusCode DelphesSimulation::finalize() {
+              << " M: "        << setw(10) << jet.Core().P4.Mass << std::endl;*/
+
+    // Reference to MC - Delphes holds references to all objects related to the Jet object,
+    // several relations might exist -> find "recursively" in a tree history a MC particle
+    std::set<int> idRefMCPart; // Avoid double counting when referencingh MC particles
+
+    Candidate* refCand  = nullptr;
+    Candidate* ref2Cand = nullptr;
+    Candidate* ref3Cand = nullptr;
+    Candidate* ref4Cand = nullptr;
+
+    TIter it1(cand->GetCandidates());
+    it1.Reset();
+    while((refCand = static_cast<Candidate*>(it1.Next()))) {
+
+      TIter it2(refCand->GetCandidates());
+      it2.Reset();
+      while((ref2Cand = static_cast<Candidate*>(it2.Next()))) {
+
+        int id = ref2Cand->GetUniqueID()-1;
+        if (id<colMCParticles->size()) idRefMCPart.insert(id);
+        else {
+
+          TIter it3(ref2Cand->GetCandidates());
+          it3.Reset();
+          while((ref3Cand = static_cast<Candidate*>(it3.Next()))) {
+
+            int id = ref3Cand->GetUniqueID()-1;
+            if (id<colMCParticles->size()) idRefMCPart.insert(id);
+            else {
+
+              TIter it4(ref3Cand->GetCandidates());
+              it4.Reset();
+              while((ref4Cand = static_cast<Candidate*>(it4.Next()))) {
+
+                int id = ref4Cand->GetUniqueID()-1;
+                if (id<colMCParticles->size()) idRefMCPart.insert(id);
+                else {
+
+                  warning() << "Can't build one of the relations from Jet to MC particle!" << std::endl;
+                }
+              } // Iter4
+            }
+          } // Iter3
+        }
+      } // Iter2
+    } // Iter1
+
+    // Test print
+    /*double totSimE = 0;*/
+
+    for (auto id : idRefMCPart) {
+
+      auto relation = ascColJetsToMC->create();
+      relation.Jet(jet);
+      relation.Particle(colMCParticles->at(id));
+
+      // Test print
+      /*double recE   = sqrt(relation.Jet().Core().P4.Px*relation.Jet().Core().P4.Px +
+                           relation.Jet().Core().P4.Py*relation.Jet().Core().P4.Py +
+                           relation.Jet().Core().P4.Pz*relation.Jet().Core().P4.Pz +
+                           relation.Jet().Core().P4.Mass*relation.Jet().Core().P4.Mass);
+      double simE   = sqrt(relation.Particle().Core().P4.Px*relation.Particle().Core().P4.Px +
+                           relation.Particle().Core().P4.Py*relation.Particle().Core().P4.Py +
+                           relation.Particle().Core().P4.Pz*relation.Particle().Core().P4.Pz +
+                           relation.Particle().Core().P4.Mass*relation.Particle().Core().P4.Mass);
+      totSimE += simE;
+      std::cout << " RefId: "    << setw(3)  << id+1
+                << " Rel E: "    << setw(10) << totSimE << " " << simE << " <-> " << recE << std::endl;*/
+    }
+    // Test print
+    /*std::cout << std::endl;*/
+
+    if (bareJet!=nullptr) delete bareJet;
```

Comment:
Thanks for the comment. I still don't get what is wrong here. Usage of raw pointers is wrong or something more deep?